tempesta_fw/str_avx2.S

```diff
@@ -0,0 +1,2296 @@
+	vpor	%xmm1, %xmm0, %xmm0
+	vmovups	%xmm0, (%rdi,%r11)
+	movq	%rbx, %r11
+	jmp	.str2low_test_len8
+
+.str2low_tolower8:
+	movq	(%rsi,%r11), %rax
+	movq	%r9, -88(%rbp)
+	movq	-88(%rbp), %mm2
+	movq	%r8, -88(%rbp)
+	movq	-88(%rbp), %mm1
+	movq	%rax, -72(%rbp)
+	movq	-72(%rbp), %mm0
+	psubb	%mm2, %mm0
+	movq	%mm0, -64(%rbp)
+	movq	-64(%rbp), %mm0
+	pcmpgtb	%mm0, %mm1
+	movq	%mm1, -88(%rbp)
+	movq	-88(%rbp), %rax
+	movq	%rax, -56(%rbp)
+	movq	-72(%rbp), %r12
+	movq	-56(%rbp), %rax
+	andq	%rcx, %rax
+	orq	%r12, %rax
+	movq	%rax, (%rdi,%r11)
+	movslq	%ebx, %rax
+	jmp	.str2low_small_len
+
+ENDPROC(__tfw_strtolower_avx2)
+
+/**
+ * Case insensitive comprison of the strings %RDI and %RSI of length %RDX.
+ * This function converts both the strings to lower case.
+ * Return 0 if strings match and non-zero otherwise.
+ * See the benchmark mentioned above for C implementation.
+ */
+ENTRY(__tfw_stricmp_avx2)
+	cmpq	$8, %rdx
+	ja	.stricmp_short
+
+	/* Process short strings below 8 bytes in length. */
+	movq	.stricmp_switch(,%rdx,8), %rax
+	jmpq	*%rax
+.section .rodata
+.align 8
+.stricmp_switch:
+	.quad	.stricmp_len0
+	.quad	.stricmp_len1
+	.quad	.stricmp_len2
+	.quad	.stricmp_len3
+	.quad	.stricmp_len4
+	.quad	.stricmp_len5
+	.quad	.stricmp_len6
+	.quad	.stricmp_len7
+	.quad	.stricmp_len8
+.text
+.stricmp_len7:
+	xorl	%eax, %eax
+	/*
+	 * The blocks at the below use complex mixture for the registers,
+	 * so leave them as they are, w/o macro generalization.
+	 */
+.stricmp_do_len7:
+	movzbl	6(%rdi), %edx
+	movzbl	6(%rsi), %ecx
+	movzbl	__tfw_lct(%rdx), %edx
+	xorb	__tfw_lct(%rcx), %dl
+	movzbl	%dl, %edx
+	orl	%eax, %edx
+.stricmp_do_len6:
+	movzbl	5(%rdi), %ecx
+	movzbl	5(%rsi), %eax
+	movzbl	__tfw_lct(%rcx), %ecx
+	xorb	__tfw_lct(%rax), %cl
+	movzbl	%cl, %eax
+	orl	%eax, %edx
+.stricmp_do_len5:
+	movzbl	4(%rdi), %eax
+	movzbl	4(%rsi), %ecx
+	movzbl	__tfw_lct(%rax), %eax
+	xorb	__tfw_lct(%rcx), %al
+	movzbl	%al, %eax
+	orl	%edx, %eax
+.stricmp_do_len4:
+	movzbl	3(%rdi), %ecx
+	movzbl	3(%rsi), %edx
+	movzbl	__tfw_lct(%rcx), %ecx
+	xorb	__tfw_lct(%rdx), %cl
+	movzbl	%cl, %edx
+	orl	%edx, %eax
+.stricmp_do_len3:
+	movzbl	2(%rdi), %edx
+	movzbl	2(%rsi), %ecx
+	movzbl	__tfw_lct(%rdx), %edx
+	xorb	__tfw_lct(%rcx), %dl
+	movzbl	%dl, %edx
+	orl	%eax, %edx
+.stricmp_do_len2:
+	movzbl	1(%rdi), %ecx
+	movzbl	1(%rsi), %eax
+	movzbl	__tfw_lct(%rcx), %ecx
+	xorb	__tfw_lct(%rax), %cl
+	movzbl	%cl, %eax
+	orl	%eax, %edx
+.stricmp_do_len1:
+	movzbl	(%rdi), %eax
+	movzbl	(%rsi), %ecx
+	movzbl	__tfw_lct(%rax), %eax
+	xorb	__tfw_lct(%rcx), %al
+	movzbl	%al, %eax
+	orl	%edx, %eax
+	ret
+.stricmp_len0:
+	xorl	%eax, %eax
+	ret
+.stricmp_len1:
+	xorl	%edx, %edx
+	jmp	.stricmp_do_len1
+.stricmp_len2:
+	xorl	%edx, %edx
+	jmp	.stricmp_do_len2
+.stricmp_len3:
+	xorl	%eax, %eax
+	jmp	.stricmp_do_len3
+.stricmp_len4:
+	xorl	%eax, %eax
+	jmp	.stricmp_do_len4
+.stricmp_len5:
+	xorl	%edx, %edx
+	jmp	.stricmp_do_len5
+.stricmp_len6:
+	xorl	%edx, %edx
+	jmp	.stricmp_do_len6
```

Comment:
I wonder, what's happening here? Why are `xorl %edx,%edx` are all around? It can be zeroed just before `jmpq *%rax`. That will remove at least one jump from cases 1 to 7. (I know it's machine-generated. I'd like to understand why compiler did that.)

Reply:
Nevermind. I now see that registers nullified are different. Sometimes it's `%rax`, not `%rdx`, because compare code alternates between them for some reason.